src/React.re

```diff
@@ -419,6 +419,76 @@ external useImperativeHandle7:
   unit =
   "useImperativeHandle";
 
+module Uncurried = {
+  [@bs.module "react"]
+  external useState:
+    ([@bs.uncurry] (unit => 'state)) =>
+    ('state, (. ('state => 'state)) => unit) =
+    "useState";
+
+  [@bs.module "react"]
+  external useReducer:
+    ((. 'state, 'action) => 'state, 'state) => ('state, (. 'action) => unit) =
+    "useReducer";
+
+  [@bs.module "react"]
+  external useReducerWithMapState:
+    (
+      [@bs.uncurry] (('state, 'action) => 'state),
+      'initialState,
+      'initialState => 'state
+    ) =>
+    ('state, (. 'action) => unit) =
+    "useReducer";
+
+  type callback('input, 'output) = (. 'input) => 'output;
+
+  [@bs.module "react"]
+  external useCallback:
+    ([@bs.uncurry] ('input => 'output)) => callback('input, 'output) =
```

Comment:
Pretty sure you want to do all of the `callback` type returns as well = https://reasonml.github.io/try?rrjsx=true&reason=C4TwDgpgBAxghgGwQIzjA1gCgOQEsA0U2A9gJRQC8RulAfEcQNwBQzA2gALIDOAdALbEAJgFcE0AEQAnCGmASAusyhQIAD2AQpAO0RQR3CAGFEKNOgBcylVEycevEdpgipUkAtt5tYEcDoMfr7ApOQU9PBIqBg4uD5+hCRBfmHWKhIGxqbR6BIszOL+AGbExJS2YfQA3gUQ-obARsRO-lSZJlHmmJiVUAAMpCwNTS09LAC+jEA

Reply:
All of the `Uncurried.useCallback*` functions should return type `Uncurried.callback`, defined as `type callback('input, 'output) = (. 'input) => 'output;`

https://reasonml.github.io/try?rrjsx=true&reason=C4TwDgpgBAxghgGwQIzjA1gCgOQEsB2YArsADRTYD2JxwAlFALxSYB0FBtDjAfBdcFoBuAFAiA2gAFkAZ1YBbSgBMiCaACIAThDTB1AXREQAHsAib8iKERkQAwohRp0ALhFQWU2ayL4YRTU0QfRY8QhImPioaEjpuPngkVAwcThJyaMFYpncodRt7R2T0dVERNWAoADNKSiYWeKgAb1yKqFtgO2p8SuYChyTnTDhIqBGAaigARjpRDw6u32A2KAAGWZEAX1EgA

However I did just discover another bug. It doesn’t look like Reason is good at handling arity-zero uncurried functions. I’m not sure what the best way around that is, unless I’m missing something.

https://reasonml.github.io/try?rrjsx=true&reason=C4TwDgpgBAxghgGwQIzjA1gCgOQEsB2YArsADRTYD2JxwAlFALxSYB0FBtDjAfBdcFoBuAFAiA2gAFkAZ1YBbSgBMiCaACIAThDTB1AXREQAHsAib8iKERkQAwohRp0ALhFQWU2ayL4YRTU0QfRY8QhImPioaEjpuPngkVAwcThJyaMFYpncodRt7R2T0dVERNWAoADNKSiYWeKgAb1yKqFtgO2p8SuYChyTnTExGgAY6UQ8Ort9gNgmRAF9RIA

Seems to be related to https://github.com/facebook/reason/issues/2357

Reply:
Oh I was oblivious. Good call for keeping continuity with the other set.

Yah that issue provides a workaround but it is kind of annoying that this error exists. I don't think it should block this PR though - if the feature is never used then a better experience won't be prioritized.